### DISCUSSION TITLE
Conditional feedback length?

### DISCUSSION THREAD

**Discussion by sp-conway** [jspsych/jsPsych#1021] [Q&A] [answered]
Hello,

I am very new to jsPsych, and I am wondering if it is possible to generate different feedback durations in a categorization task based on the subject's answer. Specifically, if the subject gets the answer correct, I would like to make their feedback 1000 ms, but 3000 ms if their choice is incorrect. This feature is critical to my experiment, as it is part of the learning phase of a categorization study, and we hope to incentivize learning by rewarding them with less waiting when they are correct.

I tried creating a function to do so, but I ran into the issue of jsPsych needing the feedback duration prior to the trial beginning (or so I think - like I said, I am very new to this). Here is what I have so far: 
```
var learning_trials_1 = {
  timeline: [
    {
      data: {screen_id: "fixation", stimulus: "+"},
      type: 'html-keyboard-response',
      stimulus: "<div style='font-size: 60px'><b>+</b></div>",
      choices: jsPsych.NO_KEYS,
      trial_duration: 1000,
      post_trial_gap:1000
    },
    {
      data: {screen_id: "learning_trial", stimulus: 'square',
      category: jsPsych.timelineVariable('category_name'),
      actual_category: jsPsych.timelineVariable('actual_category'),
      line_pos: jsPsych.timelineVariable('line_pos')
    },
      type: 'categorize-image',
      //data: jsPsych.timelineVariable("data"),
      stimulus: jsPsych.timelineVariable('square'),
      key_answer: jsPsych.timelineVariable('category_key'),
      text_answer: jsPsych.timelineVariable('category_name'),
      choices: [74,75,76],
      correct_text: "<p>Correct, this is a member of Category %ANS%.</p>",
      incorrect_text: "<p>Incorrect, this is a member of Category %ANS%.</p>",
      prompt: "<p> To which category does this object belong?  </p>"+
      "<p> <b>J</b>, <b>K</b>, or <b>L</b>?",
      show_stim_with_feedback: true,
      show_feedback_on_timeout: true,
      feedback_duration: function(data) {
        if(data.key_press == jsPsych.timelineVariable('category_key'))
        {
          var feedback_len = 1000
        }
        else{
          var feedback_len = 3000
        }
        return feedback_len
      },
      trial_duration: null,
      post_trial_gap: 10,
      }
   ],
     timeline_variables:[
    // stimuli with various properties, etc..
   ],
   randomize_order: true,
};
```

**Comment by becky-gilbert** [comment]
Hi @sp-conway, great question. This is related to how the `timelineVariable` function works, and it's something that isn't very intuitive for many people. We're trying to think of ways to improve it - see #788 for some discussion of the issue.

You had exactly the right idea for how to do this, except that when you're using `timelineVariables` _inside_ of a function, then you need to include a second `true` argument. This will return the value of the timeline variable immediately, rather than a function that returns the variable. See the [second timeline variables example in the docs](https://www.jspsych.org/overview/timeline/#timeline-variables).

Inside your `feedback_duration` function, can you try including the `true` argument inside of your `jsPsych.timelineVariable` function, like this?
```javascript
feedback_duration: function(data) {
  if (data.key_press == jsPsych.timelineVariable('category_key', true)) {
    var feedback_len = 1000
  } else {
    var feedback_len = 3000
  }
  return feedback_len
},
```

**Reply by sp-conway** [reply]
I tried that, and I get the error 
Uncaught TypeError: Cannot read property 'key_press' of undefined

**Reply by becky-gilbert** [reply]
Ah sorry, I didn't notice another problem with the code. When a non-event-related parameter is a function, that function will execute immediately _before_ the trial starts. So at the point when your `feedback_duration` function runs, there's no trial data or key press because the trial hasn't started yet. The `on_finish` function is the only one that takes a data argument (because it runs at the _end_ of the trial, when the trial data exists). 

I'm afraid I don't know the best way to make this work so that the feedback duration depends on the key press. One option is to edit the plugin, but there might be other ways to do it. Maybe someone else can jump in with suggestions.

**Reply by jodeleeuw** [reply]
I think this will require editing the plugin or splitting the trial into two trials and using html-keyboard-response for both. The second trial would show the feedback.

**Reply by sp-conway** [reply]
Hi,
thank you for the suggestions. I had ended up doing what Josh ended up suggesting, and I switched to the image-keyboard-response plugin and split it into two trials.

**Comment by becky-gilbert** [comment, marked as answer]
Copying Josh's response here and marking it as the answer to make it easier for others to find:
> I think this will require editing the plugin or splitting the trial into two trials and using html-keyboard-response for both. The second trial would show the feedback.
